torch/csrc/jit/backends/backend.h

```diff
@@ -0,0 +1,397 @@
+  };
+}
+
+c10::FunctionSchema getExecuteSchema() {
+  auto any_list_ty = c10::ListType::create(c10::AnyType::get());
+  c10::Argument self("self", c10::AnyType::get());
+  c10::Argument handle("handle", c10::AnyType::get());
+  c10::Argument input("input", any_list_ty);
+  c10::Argument output("output", any_list_ty);
+  return c10::FunctionSchema(
+      "execute",
+      /*overload_name=*/"",
+      /*arguments=*/{self, handle, input},
+      /*returns=*/{output});
+}
+
+template <typename TBackendInterface>
+std::function<void(Stack&)> getExecuteFunc() {
+  return [](Stack& stack) {
+    auto args = pop(stack);
+    auto handle = pop(stack);
+    auto self = pop(stack);
+    auto backend = self.toCustomClass<TBackendInterface>();
+    auto res = backend->execute(handle, args.toList());
+    push(stack, res);
+  };
+}
+
+// Static registration API for backends.
+template <class TBackendInterface>
+class backend {
+  static_assert(
+      std::is_base_of<PyTorchBackendInterface, TBackendInterface>::value,
+      "torch::jit::backend_<T> requires T to inherit from PyTorchBackendInterface");
+  constexpr static auto kBackendsNamespace = "__backends__";
+  std::string backend_name_;
+
+ public:
+  explicit backend(const std::string& name) : backend_name_(name) {
+    static auto cls = torch::class_<TBackendInterface>(kBackendsNamespace, name)
+                          .def(torch::init<>())
+                          ._def_unboxed(
+                              "preprocess",
+                              getPreprocessFunc<TBackendInterface>(),
+                              getPreprocessSchema())
+                          ._def_unboxed(
+                              "compile",
+                              getCompileFunc<TBackendInterface>(),
+                              getCompileSchema())
+                          ._def_unboxed(
+                              "execute",
+                              getExecuteFunc<TBackendInterface>(),
+                              getExecuteSchema());
+  }
+
+  // Generates and returns a function that takes a Module and a lowering
+  // specification in the form of a dictionary. The caller is responsible for
+  // binding this into a CPython module.
+  std::function<Module(Module, py::dict)> generateToBackendFn() {
+    const c10::QualifiedName qual_backend_name(
+        {"__torch__", "torch", "classes", kBackendsNamespace, backend_name_});
+    const std::string backend_name = qual_backend_name.name();
+
+    return [=](Module orig_module, py::dict method_compile_spec) {
+      // TODO: Validate method_compile_spec.
+
+      // Clone orig_module to make sure backend transformation is
+      // functional.
+      auto cloned_module = orig_module.clone();
+
+      // Represents of a Type of Dict[str, Any].
+      auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+
+      // Generate LoweredModule.
+      Module loweredModule("torch.jit." + backend_name + "LoweredModule");
```

Comment:
[followup] This needs to generate the module in the Python compilation unit, and properly mangle the class name. Otherwise there will be very strange behavior when you try to define more than one lowered module at a time.